virtual_rainforest/models/animals/dummy_animal_module.py

```diff
@@ -78,6 +82,75 @@ class PalatableSoil:
     """The grid position of the soil pool."""
 
 
+class AnimalCommunity:
+    """This is a class for the animal community of a grid cell."""
+
+    def __init__(
+        self,
+        functional_groups: list[FunctionalGroup],
+    ) -> None:
+        """The constructor of the AnimalCommunity class."""
+        self.functional_groups = functional_groups
+        """A list of all FunctionalGroup types in the model."""
+
+        self.cohorts: dict[str, list[AnimalCohort]] = {
+            k.name: [] for k in self.functional_groups
+        }
+        """Generate a dictionary of functional groups."""
+
+    def immigrate(self, cohort: AnimalCohort, destination: AnimalCommunity) -> None:
+        """Function to move an AnimalCohort between AnimalCommunity objects.
+
+        This function should take a cohort and a destination community and then pop the
+        cohort from this community to the destination.
+
+        Args:
+            cohort: The AnimalCohort moving between AnimalCommunities.
+            destination: The AnimalCommunity the cohort is moving to.
+
+        """
+        immigrant = self.cohorts[cohort.name].pop(
+            self.cohorts[cohort.name].index(cohort)
+        )
+        # Remove target cohort from source community
+        destination.cohorts[immigrant.name].append(immigrant)
+        # Add cohort to destination community
+
+    def die_cohort(self, cohort: AnimalCohort) -> None:
+        """The function to change the cohort status from alive to dead.
+
+        Args:
+            cohort: The AnimalCohort instance that has lost all individuals.
+
+        """
+
+        if cohort.is_alive:
+            cohort.is_alive = False
+            LOGGER.info("An animal cohort has died")
+            self.cohorts[cohort.name].remove(cohort)
+        elif not cohort.is_alive:
+            LOGGER.exception("An animal cohort which is dead cannot die.")
+
+    def birth(self, cohort: AnimalCohort) -> AnimalCohort:
```

Comment:
Should this be a method of `AnimalCohort`? A cohort will have a number of individuals in a given (average) state that might be capable of reproducing and generating a new cohort. Then at the `AnimalCommunity` level, the update might do something like:

``` python
for this_name, these_cohorts in self.cohorts.items():
    # generate new cohorts for this functional group
    new_cohorts_for_name = [chrt.reproduce() for chrt in these_cohorts if chrt.can_reproduce()]
    # potentially merge those new cohorts (all age=0) if we don't care
    # (currently/ever) about maternal and paternal effects.
    self.cohorts[this_name].extend(new_cohorts_for_name)
```

I don't think there's any issue with the content of this method, but it logically feels like it should be a step further in?

Reply:
I guess, similarly, `death` feels like an `AnimalCohort` method. So then the update would be:

``` python
for this_name, these_cohorts in self.cohorts.items():
    self.cohorts[this_name] = [chrt for chrt in these_cohorts if not chrt.just_died]
```

Reply:
The order matters too! Also semelparity would need a mechanism - unless death happens before reproduction and then a semelparous cohort could set a `die_next_check` flag?